Testing-the-engine.md

```diff
@@ -0,0 +1,302 @@
+
+For testing logic within a class at a unit level, create or add to a XCTestCase
+in the XCTest project.
+
+The iOS unit testing infrastructure is split in 2 different locations. The
+`...Test.mm` files in https://github.com/flutter/engine/tree/master/shell/platform/darwin/ios
+contain the unit tests themselves. The
+https://github.com/flutter/engine/tree/master/testing/ios/IosUnitTests directory
+contains an Xcode container project to execute the test.
+
+See https://github.com/flutter/engine/blob/master/testing/ios/IosUnitTests/README.md
+for details on adding new test files.
+
+The engine repo has a unified build system to build C, C++, Objective-C,
+Objective-C++, and Java files using [GN](https://gn.googlesource.com/gn/) and
+[Ninja](https://ninja-build.org/). Since GN and Ninja has to build the C++
+dependencies that the Objective-C classes reference, the tests aren't built by
+the Xcode project in https://github.com/flutter/engine/tree/master/testing/ios/IosUnitTests.
+
+Instead, the engine provides the script:
+
+    `testing/ios/run_tests.py --type=objc`
+
+to easily build and run the XCTests.
+
+Behind the scenes, it invokes GN and Ninja to build the tests and dependencies
+into a single `.dylib`. Then it uses Xcode and the Xcode project at
+`testing/ios/IosUnitTests` to import and execute the XCTests in the `.dylib`.
+
+See [[Setting-up-the-Engine-development-environment#editor-autocomplete-support]]
+for tips on setting up C/C++/Objective-C code completion and syntax highlighting
+when working on the engine and tests.
+
+XCTests are executed during pre-submit on our CI system when submitting PRs to
+the `flutter/engine` repository.
+
+#### XCTest
+
+[XCTest](https://developer.apple.com/documentation/xctest) is the standard way
+of creating unit tests in Xcode projects. Since iOS has x86 simulators and
+since we can build x86 engines, we can execute the XCTests directly on macOS
+in a headless simulator using the real iOS SDK.
+
+#### OCMock
+
+[OCMock](https://ocmock.org) is a standard iOS testing library used to mock
+dependencies needed to construct and test interactions with your under-test
+production class.
+
+It's best practice to test only one real production class per test and
+mock all other dependencies with OCMock.
+
+The OCMock library is an available test dependency when writing XCTests.
+
+### End-to-end tests
+
+End-to-end tests exercise the entire iOS embedding with the C++ engine on
+a headless iOS simulator. It's an integration test ensuring that
+the engine as a whole on iOS is functioning correctly.
+
+The project containing the iOS end-to-end engine test is at
+https://github.com/flutter/engine/tree/master/testing/scenario_app/ios.
+
+This test project is build similarly to a normal debug Flutter app. With the
+Dart code bundled in JIT mode and the iOS part brought into Xcode with a
+`.framework` dependency on the prebuilt local engine. It's then installed and
+executed on a simulator via Xcode.
```

Comment:
~~ With the Dart code bundled in JIT mode and the iOS part brought into Xcode with a
`.framework` dependency on the prebuilt local engine. It's then installed and executed on a simulator via Xcode.~~
The Dart code is bundled in JIT mode and is brought into the Xcode project along with a `.framework` dependency on the prebuilt local engine.  The built app is then installed and executed on a simulator via Xcode.